hardware_interface/src/resource_manager.cpp

```diff
@@ -0,0 +1,248 @@
+}  // namespace
+
+namespace resource_manager
+{
+
+ResourceManager::ResourceManager()
+{
+  actuator_loader_.reset(
+    new pluginlib::ClassLoader<hardware_interface::ActuatorHardwareInterface>(
+      "hardware_interface", "hardware_interface::ActuatorHardwareInterface"));
+  sensor_loader_.reset(
+    new pluginlib::ClassLoader<hardware_interface::SensorHardwareInterface>(
+      "hardware_interface", "hardware_interface::SensorHardwareInterface"));
+  system_loader_.reset(
+    new pluginlib::ClassLoader<hardware_interface::SystemHardwareInterface>(
+      "hardware_interface", "hardware_interface::SystemHardwareInterface"));
+
+  joint_loader_.reset(
+    new pluginlib::ClassLoader<hardware_interface::components::Joint>(
+      "hardware_interface", "hardware_interface::components::Joint"));
+}
+
+//  No real-time safe functions
+return_type
+ResourceManager::load_and_configure_resources_from_urdf(std::string urdf_string)
+{
+  std::vector<hardware_interface::HardwareInfo> hardware_info_list =
+    hardware_interface::parse_control_resources_from_urdf(urdf_string);
+
+  return_type ret = return_type::OK;
+  for (auto hardware_info : hardware_info_list) {
+    RCLCPP_INFO(
+      rclcpp::get_logger(kLoggerName),
+      "Loading hardware plugin: " + hardware_info.hardware_class_type);
+    if (!hardware_info.type.compare("system")) {
+      // TODO(anyone): this here is really not nice...
+      std::unique_ptr<hardware_interface::SystemHardwareInterface> sys_hw_if;
+      sys_hw_if.reset(system_loader_->createUnmanagedInstance(hardware_info.hardware_class_type));
+      std::shared_ptr<hardware_interface::SystemHardware> system_hw =
+        std::make_shared<hardware_interface::SystemHardware>(std::move(sys_hw_if));
+      ret = system_hw->configure(hardware_info);
+      if (ret != return_type::OK) {
+        return ret;
+      }
+      systems_.push_back(system_hw);
+      // TODO(anyone): do the same for loading sensors and actuators hardware
+    } else {
+      RCLCPP_FATAL(
+        rclcpp::get_logger(kLoggerName),
+        "hardware type not recognized");
+      return return_type::ERROR;
+    }
+
+    std::vector<std::shared_ptr<hardware_interface::components::Joint>> joints;
+    for (auto joint_info : hardware_info.joints) {
+      RCLCPP_INFO(
+        rclcpp::get_logger(kLoggerName),
+        "Loading joint plugin: " + joint_info.class_type);
+      std::shared_ptr<hardware_interface::components::Joint> joint =
+        joint_loader_->createSharedInstance(joint_info.class_type);
+      ret = joint->configure(joint_info);
+      if (ret != hardware_interface::return_type::OK) {
+        return ret;
+      }
+      joints.push_back(joint);
+      command_interfaces_[joint_info.name] = joint->get_command_interfaces();
+      // TODO(anyone): add checking if Joint has state interfaces at all
+      state_interfaces_[joint_info.name] = joint->get_state_interfaces();
+      joint_components_[joint_info.name] = joint;
```

Comment:
why do we need this? I don't really see a strong need to know that mapping. But even if, why not just storing the instance of "hardware_info_list" ?

Reply:
The controllers are accessing the joints using their names so using the names instead of `hardware_info` cannot be directly searched.